### ISSUE TITLE
http 100 continue is delayed when sending from backend to the client in version 2.2.2

### ISSUE THREAD

**Issue by MintMana** [haproxy/haproxy#825]
## Detailed description of the problem
When a client send an http post request with the header of "Expect: 100-continue" to haproxy, haproxy would delay the response 
of "100 continue" from the backend server. The delay seems to be around 200ms as described in the documentation of "option http-no-delay". However, this problem doesn't occur in haproxy 2.0.15 with the same configuration.

## Expected behavior
"http 100 continue" should be sent to the client without delay even if "option http-no-delay" is not set.

## Steps to reproduce the behavior

client: ip: 10.0.0.1 
haproxy: ip: 10.0.0.2
nginx ip : 10.0.0.2

1. The client send a POST request by curl 10.0.0.2 -XPOST -d 'any content longer than 1024 bytes to trigger the header of "Expect: 100-continue' to the haproxy server.
2. Perform a packet capture between haproxy and the client.
3. The packet capture would show an average delay around 200ms for the response of "100 continue".

## Do you have any idea what may have caused this?
No.

## Do you have an idea how to solve the issue?
Adding "option http-no-delay" to the configuration of haproxy may temporarily resolved this problem, however, it may bring other negative effects. 

## What is your configuration?
```
global
    master-worker
    nbthread 8
    chroot /var/lib/haproxy
    maxconn 80000
    user haproxy
    group haproxy
    daemon
    log 127.0.0.1 local0
    hard-stop-after 900s

defaults
    mode http
    option http-keep-alive
    log global
    timeout connect 4s
    timeout client 12s
    timeout server 60s
    http-reuse always

frontend  http_frontend
    bind *:80
    option httplog
    option dontlog-normal
    option forwardfor
    http-request del-header Connection
    default_backend http_backend

backend http_backend
    balance  roundrobin
    server  test 127.0.0.1:8081 check inter 2000 rise 2 fall 3 weight 30
```

```
worker_processes  1;

events {
    worker_connections  1024;
}


http {
    include       mime.types;
    default_type  application/octet-stream;
    access_log  logs/access.log  main;

    sendfile        on;

    keepalive_timeout  65;

server {
    listen      *:8081;
    server_name default;
    location ~ /* {
        return 200;
    }
}

```
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.2.2 2020/07/31 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.2.html
Running on: Linux 4.18.0-193.6.3.el8_2.x86_64 #1 SMP Wed Jun 10 11:09:32 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1c FIPS  28 May 2019
Running on OpenSSL version : OpenSSL 1.1.1c FIPS  28 May 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.4
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.3.1 20191121 (Red Hat 8.3.1-5)
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
	[CACHE] cache
	[FCGI] fcgi-app

```

```
4.18.0-193.14.2.el8_2.x86_64
```

## If HAProxy crashed: Last outputs and backtraces

```
(paste your output here)
```

**Comment by wtarreau** [comment]
I'm pretty sure we noticed this a few weeks ago and fixed it but I can't figure the patch. So maybe we noticed it along a code review and didn't write the fix :-(  The issue is related to MSG_MORE being still present when sending 100-continue. I'll try to reproduce.

**Comment by wtarreau** [comment]
Ah I can find the fix, so it was definitely merged: 9e3dc8305. It was backported to 2.2 as e4cfc6d1 and was merged into 2.2.1. So maybe there's a corner case :-/

**Comment by MintMana** [comment]
@wtarreau Thanks for your detailed explanation. Hope that it would be fixed soon.
The delay could be a disaster for blocking i/o http clients with only one thread and result in just 3QPS when taking network latency in consideration.

**Comment by capflam** [comment]
I pushed a patch to fix this bug. I forgot to mention the issue in the commit message. So for the record, It was fixed by the commit 7d518454b. The fix was backported to 2.2.
